### PR TITLE
Add CertificateId and VoteId data types

### PR DESCRIPTION
**Pull request by intricate**
For https://github.com/input-output-hk/ouroboros-network/issues/592 (PR https://github.com/input-output-hk/ouroboros-network/pull/999), each kind of generalized transaction has an identifier (for which we are using a hash of the generalized tx). I think it makes sense that these data types will be maintained by and exposed as part of `cardano-ledger`. Thoughts?

Also, I'm not really sure if I like the suffix, `*Id`, in this case. I was thinking that `CertificateHash` and `VoteHash` were nicer but I don't want to make things inconsistent as we already have `TxId` and `UpId`.

### CODE REVIEW

File: cardano-ledger/src/Cardano/Chain/Delegation/Certificate.hs [hunk 1]
```diff
@@ -130,6 +137,9 @@ unsafeCertificate e = UnsafeACertificate (Annotated e ())
 epoch :: ACertificate a -> EpochNumber
 epoch = unAnnotated . aEpoch
 
+certificateId :: ACertificate a -> CertificateId
+certificateId = hash . void
```
Comment (mrBliss) [inline cardano-ledger/src/Cardano/Chain/Delegation/Certificate.hs:148]: If we have a `ACertificate ByteString`, it would be more efficient to use `hashDecoded` to avoid reserialising, right? I'm basing this on `recoverUpId`. I don't know what the convention is, but maybe we could have this `certificateId` *and* `recoverCertificateId` (which uses `hashDecoded`)?

The same applies to `VoteId`.

Comment (intricate) [reply cardano-ledger/src/Cardano/Chain/Delegation/Certificate.hs:148]: To utilize `hashDecoded`, I need to implement a `Decoded` instance of `ACertificate ByteString` and, for its definition of `recoverBytes`, I probably need to add a field to store the `annotation` in `ACertificate a`. I was purposely avoiding this for now as Ru told me that annotations were being reworked in this codebase.

Comment (mrBliss) [reply cardano-ledger/src/Cardano/Chain/Delegation/Certificate.hs:148]: Ah, right. Best to make a ticket reminding us to take advantage of the reworked annotations when finished. Ideally, the ticket refers to the relevant issue/PR in `cardano-ledger` and to the locations where we should take advantage of it (looking for `void`/`hash`/`serialise` should cover most cases).

Comment (intricate) [reply cardano-ledger/src/Cardano/Chain/Delegation/Certificate.hs:148]: Okay so I just added two new functions: `recoverCertificateId` and `recoverVoteId`. I added an `annotation` field and also implemented a `Decoded` instance for both `ACertificate` and `AVote`.  I figure it made sense to go ahead and just do this now and deal with the fallout as part of the new annotations implementation.

File: cardano-ledger/src/Cardano/Chain/Delegation/Certificate.hs [hunk 2]
```diff
@@ -8,6 +8,7 @@
 {-# LANGUAGE NamedFieldPuns        #-}
 {-# LANGUAGE OverloadedStrings     #-}
 {-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
```
Comment (mrBliss) [inline cardano-ledger/src/Cardano/Chain/Delegation/Certificate.hs:11]: I'm curious (and too lazy to try it out locally :slightly_smiling_face:) why is this needed?

**Review by mrBliss — APPROVED** [review]
LGTM

File: cardano-ledger/src/Cardano/Chain/Delegation/Certificate.hs (hunk 2 above)
Comment (intricate) [reply cardano-ledger/src/Cardano/Chain/Delegation/Certificate.hs:11]: It's needed for this line in particular: 

https://github.com/input-output-hk/cardano-ledger/pull/653/files#diff-7d7801119305711a2a81d77bb358e2b9R202

Comment (mrBliss) [reply cardano-ledger/src/Cardano/Chain/Delegation/Certificate.hs:11]: Ah, of course
